docs/flatcar-linux/kvm-libvirt.md

```diff
@@ -0,0 +1,297 @@
+
+## Requirements
+
+* free RAM (minimum 4 GB)
+* libvirtd and a user who is allowed to create virtual machines
+* Terraform v0.11.x, [terraform-provider-ct](https://github.com/poseidon/terraform-provider-ct), and [terraform-provider-libvirt](https://github.com/dmacvicar/terraform-provider-libvirt) installed locally
+
+## libvirtd and User Setup
+
+*Installation on your platform:* Since we will build the terraform provider, not only libvirt should be installed but also the development packets.
+For example in Fedora it would be `sudo dnf install libvirt-daemon libvirt-devel`.
+
+Ensure that libvirt is set up and accessible for the current user:
+
+```sh
+$ sudo systemctl start libvirtd
+$ sudo usermod -a -G libvirt $(whoami)
+$ newgrp libvirt
+```
+
+## Terraform Setup
+
+Install [Terraform](https://www.terraform.io/downloads.html) v0.11.x on your system.
+
+```sh
+$ terraform version
+Terraform v0.11.13
+```
+
+Add the [terraform-provider-ct](https://github.com/poseidon/terraform-provider-ct) plugin binary for your system to `~/.terraform.d/plugins/`, noting the final name.
+
+```sh
+wget https://github.com/poseidon/terraform-provider-ct/releases/download/v0.3.1/terraform-provider-ct-v0.3.1-linux-amd64.tar.gz
+tar xzf terraform-provider-ct-v0.3.1-linux-amd64.tar.gz
+mv terraform-provider-ct-v0.3.1-linux-amd64/terraform-provider-ct ~/.terraform.d/plugins/terraform-provider-ct_v0.3.1
+```
+
+Build terraform-provider-libvirt from source because pool definition feature was not released yet:
+
+```sh
+$ go get github.com/dmacvicar/terraform-provider-libvirt
+$ cd $GOPATH/src/github.com/dmacvicar/terraform-provider-libvirt
+$ GO111MODULE=on GOFLAGS=-mod=vendor make install
+$ cp $GOPATH/bin/terraform-provider-libvirt ~/.terraform.d/plugins/terraform-provider-libvirt
+```
+
+Read [concepts](/docs/architecture/concepts.md) to learn about Terraform, modules, and organizing resources. Change to your infrastructure repository (e.g. `infra`).
+
+```sh
+cd infra/vmcluster
+```
+
+## Provider
+
+Create the `provider.tf` file:
+
+```tf
+provider "ct" {
+  version = "0.3.1"
+}
+
+provider "local" {
+  version = "~> 1.0"
+  alias   = "default"
+}
+
+provider "null" {
+  version = "~> 1.0"
+  alias   = "default"
+}
+
+provider "template" {
+  version = "~> 1.0"
+  alias   = "default"
+}
+
+provider "tls" {
+  version = "~> 1.0"
+  alias   = "default"
+}
+
+provider "libvirt" {
+  # TODO: add version once released
+  uri = "qemu:///system"
+  alias = "default"
+}
+```
+
+## Cluster
+
+Define a Kubernetes cluster using the module [kvm-libvirt/flatcar-linux/kubernetes](https://github.com/kinvolk/lokomotive-kubernetes/tree/master/kvm-libvirt/flatcar-linux/kubernetes).
+
+Create the `mycluster.tf` file:
+
+```tf
+module "controller" {
+  source = "git::https://github.com/kinvolk/lokomotive-kubernetes//kvm-libvirt/flatcar-linux/kubernetes?ref=<hash>"
```

Comment:
`ref=<hash>` will not work for users. They would have to edit it, which assumes they know about how Terraform modules work as well as the correct ref to use. Can we change this to a value that will work as is?

Reply:
I thought about adding the commit ID of the PR but that's a cyclic problem. So I just assume master will always work and added a note to add a specific commit if reproducability is needed.